pkg/object/meshcontroller/master/registrysyncer.go

```diff
@@ -136,6 +138,70 @@ func (rs *registrySyncer) handleEvent(event *serviceregistry.RegistryEvent) {
 	}
 }
 
+func (rs *registrySyncer) updateServiceSpec() {
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Errorf("recover from %v", r)
+		}
+	}()
+
+	allInstances := rs.service.ListAllServiceInstanceSpecs()
+	externalServices := make(map[string]struct{})
+	for _, instance := range allInstances {
+		if instance.RegistryName != rs.externalRegistryName() {
+			continue
+		}
+		externalServices[instance.ServiceName] = struct{}{}
+	}
+
+	allServices := rs.service.ListServiceSpecs()
+	for _, service := range allServices {
+		if service.RegistryName != rs.externalRegistryName() {
+			continue
+		}
+
+		_, exists := externalServices[service.Name]
+		if !exists {
+			logger.Infof("delete external service spec %s", service.Name)
+			rs.service.DeleteServiceSpec(service.Name)
+		}
+	}
+
+	if len(externalServices) == 0 {
+		rs.service.DeleteTenantSpec(rs.externalRegistryName())
+		return
+	}
+
+	servicesSlice := make([]string, 0)
+	for service := range externalServices {
+		servicesSlice = append(servicesSlice, service)
+	}
+	sort.Strings(servicesSlice)
+	rs.service.PutTenantSpec(&spec.Tenant{
+		Name:        rs.externalRegistryName(),
+		Services:    servicesSlice,
+		CreatedAt:   time.Now().Format(time.RFC3339),
+		Description: "Generated by registry syncer",
+	})
+
+	for service := range externalServices {
+		serviceSpec := &spec.Service{
+			RegistryName:   rs.externalRegistryName(),
+			Name:           service,
+			RegisterTenant: rs.externalRegistryName(),
+			Sidecar: &spec.Sidecar{
+				DiscoveryType:   rs.spec.RegistryType,
+				Address:         "127.0.0.1",
+				IngressPort:     13001,
+				IngressProtocol: "http",
+				EgressPort:      13002,
```

Comment:
`IngressPort` & `EgressPort` are hardcoded to 13001 & 13002, just to confirm this is desired.